src/Symfony/Component/DomCrawler/Crawler.php

```diff
@@ -553,18 +553,28 @@ public function text(string $default = null, bool $normalizeWhitespace = true):
         $text = $this->getNode(0)->nodeValue;
 
         if ($normalizeWhitespace) {
-            return trim(preg_replace("/(?:[ \n\r\t\x0C]{2,}+|[\n\r\t\x0C])/", ' ', $text), " \n\r\t\x0C");
+            return $this->normalizeWhitespace($text);
         }
 
         return $text;
     }
 
     /**
      * Returns only the inner text that is the direct descendent of the current node, excluding any child nodes.
+     *
+     * @param bool $normalizeWhitespace parameter not implemented yet
      */
-    public function innerText(): string
+    public function innerText(/* bool $normalizeWhitespace = true */): string
     {
-        return $this->filterXPath('.//text()')->text();
+        $normalizeWhitespace = 1 <= \func_num_args() ? func_get_arg(0) : true;
+
+        foreach ($this->getNode(0)->childNodes as $childNode) {
+            if (\XML_TEXT_NODE === $childNode->nodeType && '' !== trim($childNode->nodeValue)) {
+                return $normalizeWhitespace ? $this->normalizeWhitespace($childNode->nodeValue) : $childNode->nodeValue;
+            }
+        }
+
+        return '';
```

Comment:
`return $normalizeWhitespace ? '' : $this->filterXPath('.//text()')->text(false);`?

Reply:
I'd leave the `return '';`, if that's OK. Your suggestion would return `Child text` for example, for this case: `<div class="element"><span>Child text</span>  <span>Another child</span></div>` which is text from within a child node, which contradicts the methods description imo.

I've also added two more test cases where the selected element has text only within child nodes: https://github.com/symfony/symfony/pull/48940/commits/83918eb8752ba86a76c34d4ecae52e525aa97e92#diff-efbc92395bfec1b1b53ab864d9fad40ef9bfa84e08591e56ae38b3be2e83f50aR351 

I hope it's OK how I've changed it to use a dataprovider.

Reply:
When $normalizeWhitespace is false, I think we should return the untrimmed inner text, that's why I made this suggestion. Or should we remove the trim when comparing in the loop in this case?

Reply:
OK, how about a second loop without the trim?

```php
foreach ($this->getNode(0)->childNodes as $childNode) {
    if (\XML_TEXT_NODE === $childNode->nodeType && '' !== trim($childNode->nodeValue)) {
        return $normalizeWhitespace ? $this->normalizeWhitespace($childNode->nodeValue) : $childNode->nodeValue;
    }
}

if (!$normalizeWhitespace) {
    foreach ($this->getNode(0)->childNodes as $childNode) {
        if (\XML_TEXT_NODE === $childNode->nodeType) {
            return $childNode->nodeValue;
        }
    }
}

return '';
```

Removing the trim in the first loop would get you just a space character in a case like this:
```html
<div class="element"> <span>Child</span> Parent text </div>
```

I think, even with the normalize whitespace parameter set to false, this shouldn't be the result.
Should I add a second loop?

Reply:
if one opts out from whitespace normalization, it means whitespaces have a meaning for them, so yes, the space should be returned.

Reply:
OK 👍🏻, I've added the second loop: https://github.com/symfony/symfony/pull/48940/commits/f8b1d53a79b7ba44edd719ae812c77ca998a121c

Reply:
thanks, I've updated the code to really return the first non-empty text, whitespaces included.

Reply:
As the author of the original implementation, I would generally agree that your interpretation of the method description is correct, however it should be noted this is nevertheless a breaking change.